tower-http/src/cors.rs

```diff
@@ -0,0 +1,718 @@
+        self.map_layer(|layer| layer.allow_credentials(allow_credentials))
+    }
+
+    /// Set the value of the [`Access-Control-Allow-Headers`][mdn] header.
+    ///
+    /// See [`CorsLayer::allow_headers`] for more details.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Headers
+    pub fn allow_headers<I>(self, headers: I) -> Self
+    where
+        I: IntoIterator<Item = HeaderValue>,
+    {
+        self.map_layer(|layer| layer.allow_headers(headers))
+    }
+
+    /// Set the value of the [`Access-Control-Max-Age`][mdn] header.
+    ///
+    /// See [`CorsLayer::max_age`] for more details.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Max-Age
+    pub fn max_age(self, max_age: Duration) -> Self {
+        self.map_layer(|layer| layer.max_age(max_age))
+    }
+
+    /// Set the value of the [`Access-Control-Allow-Methods`][mdn] header.
+    ///
+    /// See [`CorsLayer::allow_methods`] for more details.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Methods
+    pub fn allow_methods<T>(self, methods: T) -> Self
+    where
+        T: Into<AnyOr<Vec<Method>>>,
+    {
+        self.map_layer(|layer| layer.allow_methods(methods))
+    }
+
+    /// Set the value of the [`Access-Control-Allow-Origin`][mdn] header.
+    ///
+    /// See [`CorsLayer::allow_origin`] for more details.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Allow-Origin
+    pub fn allow_origin<T>(self, origin: T) -> Self
+    where
+        T: Into<AnyOr<Origin>>,
+    {
+        self.map_layer(|layer| layer.allow_origin(origin))
+    }
+
+    /// Set the value of the [`Access-Control-Expose-Headers`][mdn] header.
+    ///
+    /// See [`CorsLayer::expose_headers`] for more details.
+    ///
+    /// [mdn]: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Access-Control-Expose-Headers
+    pub fn expose_headers<I>(self, headers: I) -> Self
+    where
+        I: Into<AnyOr<Vec<HeaderName>>>,
+    {
+        self.map_layer(|layer| layer.expose_headers(headers))
+    }
+
+    fn map_layer<F>(mut self, f: F) -> Self
+    where
+        F: FnOnce(CorsLayer) -> CorsLayer,
+    {
+        self.layer = f(self.layer);
+        self
+    }
+
+    fn is_valid_origin(&self, origin: &HeaderValue, parts: &Parts) -> bool {
+        match &self.layer.allow_origin.0 {
+            AnyOrInner::Any => true,
+            AnyOrInner::Value(allow_origin) => match &allow_origin.0 {
+                OriginInner::Exact(s) => s == origin,
+                OriginInner::List(list) => list.contains(origin),
+                OriginInner::Closure(f) => f(origin, parts),
+            },
+        }
+    }
+
+    fn is_valid_request_method(&self, method: &HeaderValue) -> bool {
+        self.layer.allow_methods == method
```

Comment:
Looks like there's a bug here. This results in a 405 method not allowed for a valid CORS request because it's comparing the list of the allowed methods directly with the requested method (`GET,POST` != `GET`).

Reply:
Ah yeah of course! Thanks for noticing. Fixed in a4f515229f447